d3a_interface/matching_algorithms/requirements_validators.py

```diff
@@ -0,0 +1,205 @@
+
+
+class EnergyTypeRequirement(Requirement):
+    """Check if energy_type requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        bid_required_energy_types = requirement.get("energy_type")
+        assert isinstance(bid_required_energy_types, list), \
+            f"Invalid data type for energy_type {requirement}"
+        offer_energy_type = (
+                offer.attributes or {}).get("energy_type", None)
+        # bid_required_energy_types is None or it is defined & includes offer_energy_type -> true
+        return not bid_required_energy_types or offer_energy_type in bid_required_energy_types
+
+
+class SelectedEnergyRequirement(Requirement):
+    """Check if energy (selected energy) requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        bid_required_energy = requirement.get("energy")
+        assert isinstance(bid_required_energy, (int, float)), \
+            f"Invalid data type for energy {requirement}"
+        assert isinstance(selected_energy, (int, float)), \
+            f"Invalid data type for selected_energy {selected_energy}"
+        # bid_required_energy is None or it is defined & equals selected_energy -> true
+        return bid_required_energy >= selected_energy
+
+
+class PriceRequirement(Requirement):
+    """Check if price requirement of bid is satisfied."""
+
+    @classmethod
+    def is_satisfied(cls, offer: Offer, bid: Bid, requirement: Dict,
+                     clearing_rate: Optional[float] = None,
+                     selected_energy: Optional[float] = None) -> bool:
+        bid_required_price = requirement.get("price")
+        assert isinstance(bid_required_price, (int, float)), \
+            f"Invalid data type for energy {requirement}"
+        assert isinstance(selected_energy, (int, float)), \
+            f"Invalid data type for selected_energy {selected_energy}"
+        assert isinstance(clearing_rate, (int, float)), \
+            f"Invalid data type for clearing_rate {clearing_rate}"
+        # bid_required_price <= price -> true
+        return bid_required_price <= selected_energy * clearing_rate
+
+
+# Supported offers/bids requirements
+# To add a new requirement create a class extending the Requirement abstract class and add it below
+
+SUPPORTED_OFFER_REQUIREMENTS = {
+    "trading_partners": TradingPartnersRequirement
+}
+
+SUPPORTED_BID_REQUIREMENTS = {
+    "trading_partners": TradingPartnersRequirement,
+    "energy_type": EnergyTypeRequirement,
+    "energy": SelectedEnergyRequirement,
+    "price": PriceRequirement
+}
+
+
+class RequirementsSatisfiedChecker:
```

Comment:
Where is this class used? I currently only see it in the myco-api-client, so I think it should be possible to move it there (without keeping it in d3a-interface). What do you think?

Reply:
It's also used in the d3a, this is why I kept it here

Reply:
https://github.com/gridsingularity/d3a/pull/1227 
I have updated the references in this PR